### ISSUE TITLE
Need to pass quotaUser parameter for use on shinyserver to stop users being disconnected

### ISSUE THREAD

**Issue by mathlete76** [tidyverse/googlesheets4#188]
I have deployed an app on shinyapps which has around 10 to 15 concurrent users throughout the day.
The app reads and writes to sheets regularly, but because I am not able to pass a quotaUser parameter for individuals the API sees all traffic from one IP and causes rate limit errors at busy times. (100 requests per 100 seconds error)

Within my app each user has a unique identifier. Is there a way to pass this as a parameter for read write operations to allow the project limits to apply?

I noticed sheets.spreadsheets.values.get$parameters$quotaUser in gs4_endpoints and wondered if that provides a method of passing a parameter.

**Comment by jennybc** [comment]
I believe I know what you're talking about, but suspect your diagnosis may be incorrect.

Can you tell me more about what sort of token you are using (oauth vs service account) and the associated GCP project?

**Comment by mathlete76** [comment]
Thanks so much for prompt reply :)

I created a service account on the google console, downloaded the json, and set options for gargle_oauth_email to the service account email, and gave gs4_auth the path to the json. On the sheet it's interacting with I have shared editing access with the service account. This worked very well with a few users, but in the past few days with more users concurrently it caused an issue.

I set up the following in the code:

serviceAccount <- 'service@MY_PROJECT.iam.gserviceaccount.com'
options(gargle_oauth_cache = ".secrets")
options(gargle_oauth_email = serviceAccount)
gs4_auth(path = '.secrets/*********.json'))

where the ******.json is the file that I downloaded from google console.

The g sheet being accessed is my solution to persistent storage of data required by the app while hosting on shinyapps.io, and is also edited by users, so every start up of the app has several read requests (20,000 read requests in total yesterday) and read/write requests are common during normal use after start up.

I may have solved the issue in a different way as I've created 10 service accounts, and on start up a user is assigned one at random along with the appropriate path to it's json. This is working well, but the app hasn't been under heavy concurrent load yet to test it fully.

**Comment by jennybc** [comment]
> I created a service account on the google console, downloaded the json, and set options for gargle_oauth_email to the service account email, and gave gs4_auth the path to the json

When you use a service account token, the `gargle_oauth_email` setting does not come into play. The only thing that matter is doing this `gs4_auth(path = '.secrets/*********.json'))`, i.e. providing the `path` to the JSON of the service account token. I recommend you delete all of these lines because they have no effect and are just confusing:

```
serviceAccount <- 'service@MY_PROJECT.iam.gserviceaccount.com'
options(gargle_oauth_cache = ".secrets")
options(gargle_oauth_email = serviceAccount)
```

The quota problems you're seeing are real. Having several service accounts, as your doing, is a reasonable solution. Or I suppose you could pay to have a higher quota and use one service. Now you're bumping up against properties of the Sheets API, not googlesheets4 or gargle.

**Comment by jennybc** [comment]
Also, depending on the quota you're hitting, all the service accounts could be using the same quota if they all belong to the same GCP project. It's possible they are only "worth it" if they are from unique GCP projects.

**Comment by jennybc** [comment]
Well, I and others regularly hit the "read requests per 100 seconds" for the GCP project that holds the default oauth creds, because that quota is effectively shared by most users of googlesheets4.

But the fact that you're using a service account at all gets you and your app users out of that pool entirely.

**Comment by mathlete76** [comment]
Just to update before I close this issue off. Using multiple service accounts has worked very well, and by setting permissions for the folder to allow all of the accounts editor access, any sheets I create in there also get the same permissions allowing all the service accounts access which has saved a bit of time.

Another part of the solution which has helped was to reduce the number off read requests. Various dataframes were stored as sheets on the same spreadsheet file, so to load them all back into the app required multiple read_sheet requests. To get around this I switched to using "drive_download" from the googledrive package to download the entire spreadsheet and have it saved as an xlsx file, and then use "read_excel" from the readxl package to grab each sheet from the downloaded file thus reducing what was 8 to 10 read requests, down to one download request per session.

read_sheet and write_sheet are still in use once the app is initialised for users to update and edit their own individual data and uses the random service account method I had implemented already.
